students/mark-l-taylor/lesson03/mailroom.py

```diff
@@ -0,0 +1,123 @@
+            'Your donation will continue to allow us to put a smile on our patients faces.', '',
+            'Sincerly,',
+            'The Last Laugh Program'])
+    print(email)
+
+def send_thanks():
+    '''Generate a thank you note. '''
+    #Prompt for name to send thank you
+    name = input('\nWho would you like to send a thank you to?\n(Tip: type \'list\' for possible names)\n')
+    if name == 'list':
+        print('Current Donors:')
+        for d in get_donor_names():
+            print(d)
+    elif name in get_donor_names():
+        amount = add_donation(name)
+        generate_email(name, amount)
+    elif name.lower() == 'quit':
+        print('Returning to main menu.')
+        return
+    else:
+        print('\n{} 1is a new donor! Adding to donor database.'.format(name))
+        #Add new donor with empty donation
+        donors.append([name, []])
+        amount = add_donation(name)
+        generate_email(name, amount)        
+             
+def donation_sort(data):
+    ''' Sort the report data by the total given.'''
+    #Report data is a list: [name, total given, num donations, avg donation]
+    #Sorting on the total given
+    return data[1]
+    
+def create_report():
+    ''' Create a formatted report of the donor data.'''
+    
+    frmt_header = '{:<26}|{:^13}|{:^11}|{:>13}'
+    frmt_line   = '{:<26} ${:>11.2f} {:>11}  ${:>12.2f}'
+    print('\nDonations Summary:\n')
+    print(frmt_header.format('Donor Name', 'Total Given', 'Num Gifts', 'Average Gift'))
+    print('-'*66)
+    
+    #Convert the donor data into report form
+    fmt_data = []
+    for d in donors:
+        fmt_data.append([d[0], sum(d[1]), len(d[1]), sum(d[1])/len(d[1])])
+
+    #Sort the data by the total amount given
+    fmt_data.sort(key=donation_sort, reverse=True)
+    
+    #Print the sorted data in the report format
+    for f in fmt_data:
+        print(frmt_line.format(*f))
+    
+def prompt_actions():
+    ''' Prompt user for action they would like to perform'''
+    
+    print('\nWhat would you like to do?')
+    for i, act in enumerate(actions):
+        print(f'\t({i+1}) {act}')
+    
+    #Get user action
+    response = input('Please select an action: ')
+    while True:
+        if not response.isnumeric() or int(response) not in list(range(1, len(actions)+1)):
+            response = input(f'Select a number between 1 and {len(actions)}: ')
+        else:
+            return int(response), actions[int(response)-1]
+    
+if __name__ == '__main__':
+
+    #Begin the script by asking the user what they want to do
+    #program will continue to loop until the user selects the Quit option
+    while True:
+        resp_num, resp_str = prompt_actions()   
+        if resp_str == 'Send a Thank You':
+            send_thanks()
+        elif resp_str == 'Create a Report':
+            create_report()
+        elif resp_str == 'Quit':
+            print('Exiting Script.')
+            exit()
```

Comment:
it is best to have user select number options instead of having the user type out the string